src/main/scala/is/hail/annotations/MemoryBlock.scala

```diff
@@ -718,13 +733,60 @@ class RegionValueBuilder(region: MemoryBuffer) {
     }
   }
 
-  def addRegionValue(t: TStruct, fromRegion: MemoryBuffer, fromOff: Long) {
-    val (toT, toOff) = current()
-    assert(toT == t.fundamentalType)
+  def addField(t: TStruct, fromRegion: MemoryBuffer, fromOff: Long, i: Int) {
+    if (t.isFieldDefined(fromRegion, fromOff, i))
+      addRegionValue(t.fieldType(i), fromRegion, t.loadField(fromRegion, fromOff, i))
+    else
+      setMissing()
+  }
 
-    region.copyFrom(fromRegion, fromOff, toOff, t.byteSize)
+  def addField(t: TStruct, rv: RegionValue, i: Int) {
+    addField(t, rv.region, rv.offset, i)
+  }
 
-    fixupStruct(t.fundamentalType, toOff, fromRegion, fromOff)
+  def addElement(t: TArray, fromRegion: MemoryBuffer, fromAOff: Long, i: Int) {
+    if (t.isElementDefined(fromRegion, fromAOff, i))
+      addRegionValue(t.elementType, fromRegion,
+        t.elementOffset(fromRegion, fromAOff, i))
+    else
+      setMissing()
+  }
+
+  def addElement(t: TArray, rv: RegionValue, i: Int) {
+    addElement(t, rv.region, rv.offset, i)
+  }
+
+  def addRegionValue(t: Type, rv: RegionValue) {
+    addRegionValue(t, rv.region, rv.offset)
+  }
+
+  def addRegionValue(t: Type, fromRegion: MemoryBuffer, fromOff: Long) {
+    val toT = currentType()
+    val toOff = currentOffset()
+    assert(toT == t.fundamentalType)
+    assert(typestk.nonEmpty || toOff == start)
+
+    t.fundamentalType match {
+      case t: TStruct =>
+        region.copyFrom(fromRegion, fromOff, toOff, t.byteSize)
+        if (region.ne(fromRegion))
+          fixupStruct(t, toOff, fromRegion, fromOff)
+      case t: TArray =>
+        // FIXME shouldn't copy if in same region (set start to fromOff?)
```

Comment:
so currently if the region value is the same it'll create another copy of the array but it won't recursively copy the array elements if they're references?